File: 2.GammaCategorization/CreationalPatterns/5.Singleton/Aula07_SingletonExcercise/Aula_08_SinlgetonExercise2/C#/CentralDeAlertas.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;


namespace CentralDeAlertasNamespace
{
    public sealed class CentralDeAlertas
    {
        // 1. Instância única garantida com volatile para visibilidade entre threads
        private static CentralDeAlertas _instancia;

        // 2. Construtor privado para impedir a criação de instâncias com 'new'
        private CentralDeAlertas()
        {
            _mensagens = new ConcurrentBag<string>();
        }


        // 3. Método global de acesso à instância única (Singleton Thread-Safe)
        public static CentralDeAlertas Instancia
        {
            get
            {
                if (_instancia == null)
                {
                    lock (_lock)
                    {
                        if (_instancia == null)
                        {
                            _instancia = new CentralDeAlertas();
                        }
                    }
                }
                return _instancia;
            }
        }


        private static readonly object _lock = new object();
        
        // ConcurrentBag é thread-safe para adições e leituras simultâneas
        private readonly ConcurrentBag<string> _mensagens;

        


        public void EnviarAlerta(string orgao, string mensagem)
        {
            string alertaFormatado = $"[{orgao}] {mensagem}";
            _mensagens.Add(alertaFormatado);
        }

        public IReadOnlyCollection<string> GetAlertas()
        {
            // Retorna uma cópia para leitura segura externamente
            return new List<string>(_mensagens);
        }
    }

}

File: 2.GammaCategorization/CreationalPatterns/5.Singleton/Aula07_SingletonExcercise/Aula_08_SinlgetonExercise2/CentralDeAlertas.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

public class CentralDeAlertas {

    private final List<String> mensagens;


    // 1. Instância única garantida com volatile para visibilidade entre threads
    private static volatile CentralDeAlertas instancia;


    // 2. Construtor privado para impedir a criação de instâncias com 'new'
    private CentralDeAlertas() {
        mensagens = new CopyOnWriteArrayList<>();
    }

    // 3. Método global de acesso à instância única (Singleton Thread-Safe)
    public static CentralDeAlertas getInstancia() {
        if (instancia == null) {
            synchronized (CentralDeAlertas.class) {
                if (instancia == null) {
                    instancia = new CentralDeAlertas();
                }
            }
        }
        return instancia;
    }


    

    public void enviarAlerta(String orgao, String mensagem) {
        String alertaFormatado = "[" + orgao + "] " + mensagem;
        mensagens.add(alertaFormatado);
    }

    public List<String> getAlertas() {
        return Collections.unmodifiableList(new ArrayList<>(mensagens));
    }
}